Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Win32Application.cpp

```diff
@@ -0,0 +1,166 @@
+{
+    // Get the Hinstance
+    HINSTANCE hInstance = GetCurrentModule();
+
+    // Initialize the window class.
+    WNDCLASSEX windowClass = { 0 };
+    windowClass.cbSize = sizeof(WNDCLASSEX);
+    windowClass.style = CS_HREDRAW | CS_VREDRAW;
+    windowClass.lpfnWndProc = WindowProc;
+    windowClass.hInstance = hInstance;
+    windowClass.hCursor = LoadCursor(NULL, IDC_ARROW);
+    windowClass.lpszClassName = L"DXQuad";
+
+    auto hr = RegisterClassEx(&windowClass);
+    // Register window class
+    auto err = GetLastError();
+    if (!hr)
+    {
+        MessageBox(NULL, L"Error registering class",
+            L"Error", MB_OK | MB_ICONERROR);
+        return false;
+    }
+
+    RECT windowRect = { 0, 0, static_cast<LONG>(pSample->GetWidth()), static_cast<LONG>(pSample->GetHeight()) };
+    AdjustWindowRect(&windowRect, WS_OVERLAPPEDWINDOW, FALSE);
+
+    // Create the window and store a handle to it.
+    m_hwnd = CreateWindow(
+        windowClass.lpszClassName,
+        pSample->GetTitle(),
+        WS_OVERLAPPEDWINDOW,
+        CW_USEDEFAULT,
+        CW_USEDEFAULT,
+        windowRect.right - windowRect.left,
+        windowRect.bottom - windowRect.top,
+        nullptr,        // We have no parent window.
+        nullptr,        // We aren't using menus.
+        hInstance,
+        pSample);
+    if (!m_hwnd)
+    {
+        MessageBox(NULL, L"Error creating window",
+            L"Error", MB_OK | MB_ICONERROR);
+        return false;
+    }
+
+    // Initialize the sample and show the window
+    pSample->OnInit();
+    ShowWindow(m_hwnd, nCmdShow);
+
+    // Main sample loop.
+    close_window = false;
+    MSG msg = {};
+    while (msg.message != WM_QUIT)
+    {
+        // Destroy the sample and window if the close_window
+        // variable is set to true
+        if (close_window) {
+            pSample->OnDestroy();
+            PostMessage(m_hwnd, WM_CLOSE, 0, 0);
+            break;
+        }
+        // Process any messages in the queue.
+        if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE))
+        {
+            TranslateMessage(&msg);
+            DispatchMessage(&msg);
+        }
+    }
+
+    // Unregister the hInstance so it can be reused if
+    // the sample is reloaded
+    if (!UnregisterClass(L"DXQuad", hInstance))
+    {
+        auto error = GetLastError();
+        MessageBox(NULL, L"Error unregistering class",
+            L"Error", MB_OK | MB_ICONERROR);
+        return 1;
+    }
+    else {
+        OutputDebugString(L"Class unregistered");
+    }
+
+    // Return this part of the WM_QUIT message to Windows.
+    return static_cast<char>(msg.wParam);
+}
+
+// Main message handler for the sample.
+LRESULT CALLBACK Win32Application::WindowProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    D3D12Quad* pSample = reinterpret_cast<D3D12Quad*>(GetWindowLongPtr(hWnd, GWLP_USERDATA));
+
+    switch (message)
+    {
+    case WM_CREATE:
```

Comment:
>WM_CREATE [](http://example.com/codeflow?start=9&length=9)

WM_NCCREATE is safer here for setting the class window pointer because it's the first message you receive, but there can be other messages before WM_CREATE. (In this case, it's actually okay because we don't have a bunch of other messages to handle, but better safe than sorry and better to set right example for others)

```
    case WM_NCCREATE:
        {
            // Save the D3D12Quad* passed in to CreateWindow.
            LPCREATESTRUCT pCreateStruct = reinterpret_cast<LPCREATESTRUCT>(lParam);
            SetWindowLongPtr(hWnd, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(pCreateStruct->lpCreateParams));
            return DefWindowProc(hWnd, message, wParam, lParam);
        }
```

Note DefWindowProc should be called for WM_NCCREATE.

[https://devblogs.microsoft.com/oldnewthing/20191014-00/?p=10299](https://devblogs.microsoft.com/oldnewthing/20191014-00/?p=10299)

Reply:
Done.